internal/serverinstall/docker.go

```diff
@@ -240,6 +240,201 @@ func (i *DockerInstaller) Install(
 	}, nil
 }
 
+// Upgrade is a method of DockerInstaller and implements the Installer interface to
+// upgrade a waypoint-server as a Docker container
+func (i *DockerInstaller) Upgrade(
+	ctx context.Context, opts *InstallOpts, serverCfg serverconfig.Client) (
+	*InstallResults, error,
+) {
+	ui := opts.UI
+
+	sg := ui.StepGroup()
+	defer sg.Wait()
+
+	s := sg.Add("Initializing Docker client...")
+	defer func() { s.Abort() }()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, err
+	}
+	cli.NegotiateAPIVersion(ctx)
+
+	s.Update("Checking for an existing Waypoint server installation...")
+	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "label",
+			Value: "waypoint-type=server",
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	grpcPort := defaultGrpcPort
+	httpPort := defaultHttpPort
+
+	var (
+		clicfg   clicontext.Config
+		addr     pb.ServerConfig_AdvertiseAddr
+		httpAddr string
+	)
+
+	clicfg.Server = serverconfig.Client{
+		Address:       "localhost:" + grpcPort,
+		Tls:           true,
+		TlsSkipVerify: true,
+	}
+
+	addr.Addr = serverName + ":" + grpcPort
+	addr.Tls = true
+	addr.TlsSkipVerify = true
+
+	httpAddr = "localhost:" + httpPort
+
+	if len(containers) < 0 {
+		s.Update("No waypoint server detected. Nothing to upgrade.")
+		s.Status(terminal.StatusWarn)
+		s.Done()
+		return nil, fmt.Errorf("No waypoint server container detected")
+	}
+
+	// Assume waypoint-server is the first container with the waypoint-type label
+	waypointServerContainer := containers[0]
+
+	s.Update("Checking for Docker image: %s", i.config.serverImage)
+
+	imageRef, err := reference.ParseNormalizedNamed(i.config.serverImage)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing Docker image: %s", err)
+	}
+
+	imageList, err := cli.ImageList(ctx, types.ImageListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "reference",
+			Value: reference.FamiliarString(imageRef),
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(imageList) == 0 {
+		s = sg.Add("Pulling image: %s", i.config.serverImage)
```

Comment:
Do we have to actually always pull? Otherwise `waypoint upgrade` on Docker will reuse an old `hashicorp/waypoint:latest`, right? 

We can probably follow the K8S/Nomad logic here and only always pull if the tag is `:latest`.

Reply:
Yeah, always pulling on `:latest` here is a good idea! Good catch :+1: